### PR TITLE
Make Unicode characters printable and enterable

### PR DESCRIPTION
**Pull request by direc85**
This PR enables BusyBox to both print Unicode characters in e.g. `ls` output, and render them correctly in command input, e.g. `mv` or `cp` filenames.

Currently busybox utils like "ls" fail to display filenames containing UTF-8 characters, replacing any special characters with "?".

Change libbb's printable_string() function to allow high ASCII characters so that unicode filenames are displayed correctls.

Sailfish OS Forum thread:
https://forum.sailfishos.org/t/asian-character-support-in-terminal/12938

Cherry-picked from OpenWRT:
https://git.openwrt.org/?p=openwrt/openwrt.git;a=commit;h=2ab0dae9915d5c7c1c06c26ce1956bf0badd7a5b

Oritinal issue and changeset:
https://dev.archive.openwrt.org/ticket/7993.html
https://dev.archive.openwrt.org/changeset/43084.html

Additionally, it looks like Unicode support was partially disabled in the configurations, too. The second commit sets `CONFIG_LAST_SUPPORTED_WCHAR=0` to enable considering all printable Unicode characters, and `CONFIG_UNICODE_COMBINING_WCHARS=y` and `CONFIG_UNICODE_WIDE_WCHARS=y` are needed to make them appear on the terminal when e.g. tab-completing or pasting.

I installed all the non-conflicting RPM packages generated on my Sony Xperia 10 II, and after a successful reboot, I tested the following things:

- `vim 緑スズメバチ.txt`
- `cat 緑スズメバチ.txt`
- `ls`
- `find -name "緑スズメバチ.txt"`

I also did some typical tasks to me (editing and moving files, using `zypper` etc.) and I didn't find anything out of place in the case of "simple" UTF-8 characters (ASCII plus Finnish letters).

It should be noted that the double-wide characters somewhat break the terminal character layout, but this is the case for at least Bash and Fish shells too - I'd assume the users have correct fonts installed to make text render correctly using SSH at least. As Sailfish uses ASCII filenames and paths, at least as far as I can tell, I do not see this as a blocking issue for the existing use cases.

### CODE REVIEW

**Comment by pvuorela** [comment]
I'm puzzled that the openwrt patch is from 2014 and yet the upstream busybox still fails to show unicode filenames (right?). How's that even possible? Was the patch not offered upstream (didn't spot anything on the linked tickets) or did the upstream not want these changes.

**Comment by direc85** [comment]
I wondered the same thing. OpenWRT still has the patch, and `printable_string.c` hasn't been touched in four years, so it's not impossible that the patch simply hasn't been offered upstream. I tried to look for it, but couldn't find anything either.

Do you think it would be better to offer the patch or change upstream instead?

**Comment by pvuorela** [comment]
> Do you think it would be better to offer the patch or change upstream instead?

Regardless whether we merge this here or not, ideally the upstream should get fixes (and the patches eventually removed). Also while the diff here isn't big, it increases confidence level of merging things if the same change can be found upstream.

For the config the CONFIG_LAST_SUPPORTED_WCHAR was on the lower side so already that could explain the output problems. Did you test how this works just by the config changes and not touching the .c?

File: rpm/busybox.spec [hunk 1]
```diff
@@ -14,6 +14,7 @@ Patch2:  0003-Align-watch-with-what-is-in-procps-ng.patch
 Patch3:  0004-ash-Load-ENV-file-also-if-SSH_CLIENT-SSH2_CLIENT-is-.patch
 Patch4:  0005-ash-job-option-to-restore-term-io-after-job-is-stopp.patch
 Patch5:  0006-ash-Write-history-on-SIGHUP.patch
+Patch7:  0007-make-unicode_printable.patch
```
Comment (ighea) [inline rpm/busybox.spec:17]: Could you set the patch index to be Patch6?

**Comment by direc85** [comment]
> For the config the CONFIG_LAST_SUPPORTED_WCHAR was on the lower side so already that could explain the output problems. Did you test how this works just by the config changes and not touching the .c?

No, I did the .c patch first, and then discovered the input issues. I'll test it today and post the results.

> Regardless whether we merge this here or not, ideally the upstream should get fixes (and the patches eventually removed).

Indeed. I test this a bit more and see what upstream thinks about the .c change.

File: rpm/0007-make-unicode_printable.patch [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+ 			unsigned char c = *d;
+ 			if (c == '\0')
+ 				break;
+-			if (c < ' ' || c >= 0x7f)
++			if (c < ' ')
+ 				*d = '?';
+ 			d++;
+ 		}
+-- 
+2.37.3
```
Comment (Thaodan) [inline rpm/0007-make-unicode_printable.patch:48]: please generate patch with --no-signature to not have the trailing git version in the path.

**Comment by direc85** [comment]
Well now, it looks like the patch isn't needed at all - the config changes alone achieve the same result. I had to double check and reboot the device an extra time to be sure. I guess there's nothing to upstream then!

I'll push the removal of the patch and leave the config changes in.

File: rpm/0007-make-unicode_printable.patch (hunk 2 above)
Comment (direc85) [reply rpm/0007-make-unicode_printable.patch]: Thanks for the tip, I'll keep that in mind for any future patches!

**Comment by Thaodan** [comment]
> I'll push the removal of the patch and leave the config changes in.

Please use git rebase --interactive to remove the revert and the original commit that adds the patch.

**Comment by Thaodan** [comment]
Hey I opened a bug for this internally. Can you please follow the existing commit scheme for commits that contain bug references and add the reference JB#58783? The message should look something like: "[busybox] Enable support for Unicode characters after Latin-Extend B. Fixes JB#58783 <sign off>"


I tested the issue more thoroughly, the affected characters are all from characters that come after the Latin-Extend B.
You can test so by copying characters for a character selection tool such as Kcharselect.

**Review by pvuorela — APPROVED** [review]

**Pull request merged**